examples/seismic/acoustic/operators.py

```diff
@@ -1,7 +1,39 @@
+from sympy import sign
+
 from devito import Eq, Operator, Function, TimeFunction, Inc
+from devito.symbolics import retrieve_functions, INT
 from examples.seismic import PointSource, Receiver
 
 
+def freesurface(model, pde, u):
+    """
+    Generate the stencil that mirrors the field as a free surface modeling for
+    the acoustic wave equation
+    Parameters
+    ----------
+    field: TimeFunction or Tuple
+        Field for which to add a free surface
+    npml: int
+        Number of ABC points
+    forward: Bool
+        Whether it is forward or backward propagation (in time)
+    """
+    lhs = pde.lhs
+    rhs = pde.rhs.evaluate
+    # Add modulo replacements to to rhs
+    z = model.grid.dimensions[-1]
+    zfs = model.grid.subdomains['fsdomain'].dimensions[-1]
+
+    funcs = retrieve_functions(rhs)
+    mapper = {}
+    for f in funcs:
+        zind = f.indices[-1]
+        if (zind - z).as_coeff_Mul()[0] < 0:
+            s = sign(zind.subs({z: zfs, z.spacing: 1}))
+            mapper.update({f: s * f.subs({zind: INT(abs(zind))})})
+    return Eq(lhs, rhs.subs(mapper), subdomain=model.grid.subdomains['fsdomain'])
+
+
 def laplacian(field, m, s, kernel):
     """
     Spacial discretization for the isotropic acoustic wave equation. For a 4th
```

Comment:
Should that be spatial or am I wrong? I know both are sometimes used.

Reply:
fixed